core/src/sumeragi/main_loop.rs

```diff
@@ -837,6 +826,7 @@ pub(crate) fn run<F: FaultInjection>(
         }
 
         let mut message = sumeragi.receive_network_packet(&state, &mut view_change_proof_chain);
+        should_sleep = message.is_none();
         handle_role_agnostic_message(sumeragi, &mut state, &mut message, &mut voting_block);
```

Comment:
When refactoring I remember thinking message should never be shared as a mutable reference but instead it should be matched only once. This wouldn't have happened if this were true

Reply:
this is still very brittle IMO

Reply:
> but instead it should be matched only once

@mversic i do as you suggested by extracting message processing into it's own function. 

there is still a lot of mutable variables if future we can try to refactor as state machine

Reply:
i like it